### ISSUE TITLE
Adding support to storing Container's package name for OdspDriverUrlResolver  to enable prefetching of Container's scripts

### ISSUE THREAD

**Issue by abrahamrq** [microsoft/FluidFramework#3947]
To enable prefetching of Container's scripts in parallel of loading fluid content, we would like to add a header containing the package name. For now, the plan is to start adding this support to OdspResolvedUrl and later discuss to extend it to FluidResolvedUrl (we see that this could be used by other storage drivers to improve perf).

The way I'm thinking to add this functionality is by adding a request header to and then adding it to the resolved url so we are able to access it, similar to what we do with create new filename.

Next steps: 
After having the container's package name available for us, make the needed changes to the bohemia code loader to be able to leverage this information and make the network request before having completely loaded the fluid content (we can do this in parallel). 
We need to take into consideration the use of cached scripts, right now we have this functionalitym but this change is going to enable us to be generic and rely on package name embedded in the link (as a header)

**Comment by abrahamrq** [comment]
#3944

**Comment by anthony-murphy** [comment]
What i'm not seeing is how is this persisted, especially in the load case? where does the host issuing the request to load a container get this value to set it in the header? If the host already know it, then it seems they could just load the code, and it doesn't need to be in the resolved url?

**Comment by christiango** [comment]
Couple of thoughts: 
1. In general, I'd like to have line of sight on how much logic around URL resolving lives in the FF driver vs our internal repo? Is there a plan for convergence? I believe we already do some trickery for the component URL in the internal repo
2. In terms of how we will use the information, we can probably take a staged approach, where office-web-host uses this information to manually call preload on the code loader in parallel to the loader call. Once we refine the APIs a bit, we can form an opinion on what the ICodeLoader preload API should look like and we could move the actual preload out of office-web-host and into FF driver/loader

**Comment by christiango** [comment]
> What i'm not seeing is how is this persisted, especially in the load case? where does the host issuing the request to load a container get this value to set it in the header? If the host already know it, then it seems they could just load the code, and it doesn't need to be in the resolved url?

I have the same question, I think this has to end up in the URL. I don't have context on where the logic around encoding fluid info into the ODSP nav lives, my guess is that it is in office-web-host.

**Comment by anthony-murphy** [comment]
In addition to resolving urls the url reslover can also generate them via `getAbsoluteUrl`. I think if that function was encoding the container type in the url it would make sense, but then i don't think the load path would pass the container type as a header, as it would be expected to be encoded in the url itself.

**Comment by jatgarg** [comment]
> Couple of thoughts:
> 
> 1. In general, I'd like to have line of sight on how much logic around URL resolving lives in the FF driver vs our internal repo? Is there a plan for convergence? I believe we already do some trickery for the component URL in the internal repo
> 2. In terms of how we will use the information, we can probably take a staged approach, where office-web-host uses this information to manually call preload on the code loader in parallel to the loader call. Once we refine the APIs a bit, we can form an opinion on what the ICodeLoader preload API should look like and we could move the actual preload out of office-web-host and into FF driver/loader

So all the nav parameter encoding/sharing link generation used to live in FFX repo before 0.27. In 0.27, both the resolvers have been merged and ffx will start using the FF resolver(mostly entirely but maybe if they want to add anything on top of that then they can wrap FF resolver just like today).

**Comment by jatgarg** [comment]
I also second Tony's opinion that if the host already knows the package info(whether it adds it in url or in header), then it can do code loading stuff already. Why it needs to be added in the resolved url?

**Comment by christiango** [comment]
> I also second Tony's opinion that if the host already knows the package info(whether it adds it in url or in header), then it can do code loading stuff already. Why it needs to be added in the resolved url?

I don't think the host will know what it is, it will need to be encoded in the URL in the nav parameter and extracted from there

**Comment by abrahamrq** [comment]
> Couple of thoughts:
> 
> 1. In general, I'd like to have line of sight on how much logic around URL resolving lives in the FF driver vs our internal repo? Is there a plan for convergence? I believe we already do some trickery for the component URL in the internal repo
> 2. In terms of how we will use the information, we can probably take a staged approach, where office-web-host uses this information to manually call preload on the code loader in parallel to the loader call. Once we refine the APIs a bit, we can form an opinion on what the ICodeLoader preload API should look like and we could move the actual preload out of office-web-host and into FF driver/loader

I agree with #2, I think that is the approach we are going for. 

I added the change to include the containerpackagename in the nav parameter but are we saying that headers don't work for this kind of work because we need to be able to store it in the url?

**Comment by christiango** [comment]
Given that the container package information is going to be stored in the _real_ URL via the nav parameter, it feels like having it be a part of the URL rather than a header might be a better fit no? 

Is the goal of the V2 URL resolver that we can just pass a share URL and it would automatically know how to decode code hint out of it?

**Comment by abrahamrq** [comment]
FYI, i moved the implementation to use the URL instead of headers

**Comment by anthony-murphy** [comment]
The current change looks better, but i think there is still a bootstrapping problem. If the hint exists, it can add it to the url generated via getAbsoluteUrl, and those urls will also have the hint, which is good. But what about the case where there is no hint, like a storage generated url in something like an at mention email. Or is that not a case we are trying to solve here.

**Comment by anthony-murphy** [comment]
Before we approve any change I think we should understand and agree on the full life cycle of the hint

**Comment by abrahamrq** [comment]
> The current change looks better, but i think there is still a bootstrapping problem. If the hint exists, the resolver can add it to the url generated via getAbsoluteUrl, and those urls will also have the hint, which is good. But what about the case where there is no hint, like a storage generated url in something like an at mention email. In that case any link created here will also be missing the hint, and so on. Or is that not a case we are trying to solve here.

For now if there is no hint, we would not use the preloading functionality and fallback to the current behaviour. the plan was to add this and start making changes in the bohemia code loader. Right now Bohemia code loader works with the assumption that we always load office-fluid-container, with this change, we can extend the functionality to be generic and to rely on package name embedded in the link.  After that, we can discuss if we want to move the functionality to be in the FF side.

**Comment by anthony-murphy** [comment]
I just think we will lose the hint quickly. For instance, if someone creates a new document, creates no links, and then closes the document the hint will be lost, as it is not persisted. 

We need to at least have a plan for subsequent re-openings, and how we will rehydrate the hint, as that could have implications in the hint design.

**Comment by christiango** [comment]
> I just think we will lose the hint quickly. For instance, if someone creates a new document, creates no links, and then closes the document the hint will be lost, as it is not persisted.
> 
> We need to at least have a plan for subsequent re-openings, and how we will rehydrate the hint, as that could have implications in the hint design.

This is optimizing for the scenario where we are loading from a link, like Teams/OWA. Can you clarify what you are referring to here Tony? ODSP doc lib click or something? Do you have an alternate proposal in mind for the design here?

**Comment by anthony-murphy** [comment]
Any links created from a document without a hint won't have the hint, it is lost. So only links created at time of original create will have the hint. All subsequent opens from things like the doc link, mru, at mention emails, all other means of access will not have the hint, and links created there will not have it either.

**Comment by anthony-murphy** [comment]
Unless i'm missing something? Is there something in the PR that gets odsp to persist the container type an return it on all subsequent odsp generated links for that file?

**Comment by anthony-murphy** [comment]
My proposal would be to persist it a container create time and try to rehydrate if it exists. i don't know what specific mechanisms odsp has for this

**Comment by christiango** [comment]
> My proposal would be to persist it at container create time and try to rehydrate if it exists. i don't know what specific mechanisms odsp has for this

Can you elaborate more on what this means? Adding a network roundtrip here is not an option we want to go with, what kind of mechanism are you looking for from ODSP here?

I would expect this PR to add the ability to persist this in the ODSP URLs generated from containers for copy/paste scenarios, I'm not familiar enough with the code to know where that should happen.

**Comment by abrahamrq** [comment]
I thought that the change I made was doing that (persisting the container package in the ODSP url), is there something I'm missing?

**Comment by jatgarg** [comment]
You are just adding it to the resolved url which is always created afresh during create or load. It is not persisted anywhere. The most you can do currently is to just cache the resolved url but that is just for same host. 
So yeah is will not persist the way you want.

**Comment by abrahamrq** [comment]
> You are just adding it to the resolved url which is always created afresh during create or load. It is not persisted anywhere. The most you can do currently is to just cache the resolved url but that is just for same host.
> So yeah is will not persist the way you want.

were do we need to make the changes to be able to persist this?

**Comment by jatgarg** [comment]
> > My proposal would be to persist it at container create time and try to rehydrate if it exists. i don't know what specific mechanisms odsp has for this
> 
> Can you elaborate more on what this means? Adding a network roundtrip here is not an option we want to go with, what kind of mechanism are you looking for from ODSP here?
> 
> I would expect this PR to add the ability to persist this in the ODSP URLs generated from containers for copy/paste scenarios, I'm not familiar enough with the code to know where that should happen.

If you want to use it in copy/paste scenarios, then I think you need to add this ability in "appendDataStorePath" api. I think this is the api which is used by app resolver to generate url when copy/paste. You may need to add this hint in the url generated.

**Comment by anthony-murphy** [comment]
I think you will need to make another network round trip if there isn't a hint. At container create time the resolver should kick off a request to put the hint somewhere (odsp metadata?), probably doesn't need to block. 

On url resolve you should make a request for that metadata. To avoid perf issue i wouldn't block the resolve on this request, eventually it will show up, and you will put it somewhere. You probably want to block getAbsoluteUrl on this request, so the hint can then be appended.

**Comment by anthony-murphy** [comment]
I was thinking about this a bit more, and there may be another path. We could potentially pass the containers code details to url resolver:
``` typescript
export interface IUrlResolver {
    resolve(request: IRequest): Promise<IResolvedUrl | undefined>;

    getAbsoluteUrl(
        resolvedUrl: IResolvedUrl,
        relativeUrl: string,

        // this param would be new and optional
        codeDetails?: IFluidCodeDetails,
    ): Promise<string>;
}
```

this method is called by Container, so has easy access to them. This should also get rid of the need to have the ResolvedUrl involved. I think i'm ok with this, as the codeDetails and the url resolver are already something the host needs to be aware of, so it doesn't feel like it's introducing new concepts, or crossing our layer boundaries. I don't love the mixing of code loading and drivers, but it does make it natural to encode package details into the url in a runtime agnostic way.

Curious what @ChumpChief , @vladsud , and @curtisman  think about this idea.

**Comment by vladsud** [comment]
It sounds reasonable to me

**Comment by ram-ramasubramanian** [comment]
Catching up here. Looks like there are 2 questions:
1. Adding the container hint to the URL (works for the scenario flow) - this is critical to our shipping scenarios in Dec.
2. How do we handle other cases where a sharing link is generated from other sources (like DocLIb etc.)?

We can work with the Sharing team to see how to add hints here in general. There might be prior art we can fallback on here.

**Comment by anthony-murphy** [comment]
this was fixed by #4001
